Cluster seeding

martinfurmanski opened the pull request:
changelog: Fix cluster seeding failing in certain scenarios

RagnarW committed:
Improve ClusterSeedingIT to cover more scenarios

* Split up ClusterSeedingIT in two parameterized tests.
ClusterSeedingIT and SeedingNewMemberIt.
* Each test will seed with each type of BackupStore. EmptyBackup,
BackupContainingSomeData, BackupContainingSomeDataAndNoTransactionLogs
* Added monitors to be able to track and assert on store copy.

RagnarW committed:
Expanded seeding ITs

* Added NoStore option in ClusterSeedingIT
* Added intermediate load in NewMemberSeedingIT

martinfurmanski committed:
Fix cluster seeding

RagnarW commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/StoreCopyClient.java:
```diff
@@ -77,7 +71,7 @@ public boolean onFileContent( CompletableFuture<Long> signal, FileChunk fileChun
                                 StoreCopyFinishedResponse response )
                         {
-                            log.info( "Finished streaming %s", destination );
+                            log.info( "Finished streaming %s" );
```
this will not log destination

martinfurmanski replied:
Sloppy removal, will remove %s as well. The destination was removed on purpose because that is already logged when receiving and is not interesting to log again on completion.

RagnarW commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/RemoteStore.java:
```diff
@@ -123,20 +124,40 @@ private long getPullIndex( File storeDir ) throws IOException
+     * they end and pull from there, excluding the last one so that we do not
+     * get duplicate entries.
+     */
+    public CatchupResult tryCatchingUp( AdvertisedSocketAddress from, StoreId expectedStoreId ) throws StoreCopyFailedException, IOException
```
I think we are still of by one. Assume that i = last index, T=tx, M=metadata.

In the scenario were tx logs are present we would before see:
    CLIENT -> Ti -> SERVER
    SERVER -> Ti+1 -> CLIENT

Now we have:
  CLIENT -> Ti-i -> SERVER
  SERVER -> Ti -> CLIENT.

In summary what we get from server now is one of:
Ti
Mi or Mi+1

What we want is:
Ti+1
Mi or Mi+1 

isn't it?

RagnarW replied:
Disregard this comment. I was wrong.

andrewkerr9000 commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/LocalDatabase.java:
```diff
@@ -252,4 +253,15 @@ private void dropAvailabilityGuard()
         currentRequirement = null;
     }
+
+    public boolean hasTxLogs() throws RuntimeException
```
Does it make sense to declare `throws RuntimeException`?

martinfurmanski replied:
Kind of pointless indeed.

andrewkerr9000 reviewed (approved):
One comment, but I'm not particularly familiar with this area.

martinfurmanski committed:
Strict handling of illegal requests

The catchup client should never try to pull BASE_TX_ID (=1), so
the server side handles that in a strict manner.
